---
title: "Analyze and visualize RNA-Seq data with pathnet"
author:
- Andy An
- Travis Blimkie
output:
    BiocStyle::html_document:
        toc: true
        toc_depth: 3
        toc_float: true
vignette: >
  %\VignetteIndexEntry{Analyze and visualize RNA-Seq data with pathnet}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<head>
<link rel="shortcut icon" href="favicon_32.svg">
<link rel="shortcut icon" href="favicon_16.svg">
</head>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "")
library(dplyr)
```


# Introduction
Often times, gene expression studies such as microarrays and RNA-Seq result in
hundreds to thousands of differentially expressed genes (DEGs). It becomes very
difficult to understand the biological significance of such massive data sets,
especially when there are multiple conditions and comparisons being analyzed.
This package facilitates visualization of downstream analyses for differential
gene expression results, using pathway enrichment and protein-protein
interaction networks, to aid researchers in uncovering underlying biology and
pathophysiology from their gene expression studies.


# Installation
To install and load the package:
```{r load_library}
# Install code will go here, make sure to set "eval=FALSE"!
library(pathnet)
```


# Creating volcano plots
One of the first visualizations with gene expression studies is finding how many
DEGs there are. DEGs are usually defined by specific cutoffs for fold changes
and adjusted p-values. Generally, cutoffs of adjusted p-value <0.05 and absolute
fold change >1.5 are used, which are set as defaults here. To generate a volcano
plot, we can use the function `eruption()` in this package.

We have included example gene expression study results in this package:
`deseq_example_list`. This is a list of 6 data frames generated using DESeq2
(Love et al. 2014). Data is from an RNA-Seq study investigating COVID-19 and
non-COVID-19 sepsis patients at admission (T1) compared to ~1 week later (T2)
in the ICU, indexed over time (i.e., T2 vs T1) (An et al. 2023).
```{r datasets_and_preliminary_volcano_plot}
# Two different comparisons are included
names(deseq_example_list)

# A quick look at what a DESeq2 result looks like
deseq_example <- deseq_example_list[[1]]
as_tibble(deseq_example)

# Generate a volcano plot using this first dataframe
eruption(deseq_example)
```

There are multiple options available for customizing this volcano plot. You can
add a title, adjust cutoffs, change colours, change the number of labelled
points, change x and y axis limits, convert between log2 and non-log fold change
(for a better sense of magnitude), and highlight specific genes.

Since we're looking at COVID-19 in these example data sets, let's highlight
genes in the interferon signaling pathway. We will use a database of Reactome
pathways used by the pathway enrichment software `sigora` to generate a list of
genes involved in interferon signaling. More on Reactome and Sigora in the
pathway enrichment section below.
```{r customized_volcano_plot}
ifn_genes <- sigora_database %>% 
    filter(pathway_name == "Interferon Signaling") %>% 
    pull(Ensembl.Gene.ID)

# Here we give the following options:
# - Highlight these interferon genes
# - Give them a description on the plot
# - Converts the x-axis from log2 fold change to non-log2 fold change
# - Label only the top 5 up and down-regulated genes
# - These top genes should be only from the interferon gene list
eruption(
    deseq_results = deseq_example,
    title         = names(deseq_example_list)[1],
    select_genes  = ifn_genes, 
    select_name   = "Interferon genes", 
    nonlog2       = TRUE, 
    n             = 5, 
    label         = "select" 
)
```

We see quite a lot of interferon signaling genes that are down-regulated in 
COVID-19 patients at ICU admission, perhaps reflective of some resolution of an
antiviral response. However, it's difficult to get a sense of whether this is 
due to chance or if there is a true biological significance, as there are also 
many interferon genes that were not DEGs. This is where pathway enrichment can 
tell us statistically if what we see is real biology or not.


# Building and visualizing PPI networks
`pathnet` also provides tools for constructing and visualizing Protein-Protein
Interaction (PPI) networks. We leverage PPI data gathered from [InnateDB]() to
generate a list of interactions among DE genes identified in RNA-Seq analyses.
These interactions can then be used to build PPI networks using tools from
`igraph` and `tidygraph`, will multiple options for controlling the resulting
type of network, such as support for first, minimum, or zero order networks. The
two main functions used to accomplish this are `ppi_build_network` and
`ppi_plot_network`.
```{r ppi_networks, fig.width=12, fig.height=10, warning=FALSE}
# Start by filtered one of the example DE gene lists for significant genes
ex_de_genes <- deseq_example_list[[1]] %>%
    tibble::rownames_to_column("gene") %>%
    as_tibble() %>%
    filter(padj < 0.05 & abs(log2FoldChange) > log2(1.5))

# Construct a zero order network, i.e. it will only contain the input genes, 
# no interactors
ex_network <- ppi_build_network(
    df = ex_de_genes,
    col = "gene",
    order = "zero"
)

# Plot the PPI network, colouring the nodes based on their log2 fold change
# values. Many options are available to customize the final plot; see
# `?ppi_plot_network` for all the details.
ppi_plot_network(
    ex_network,
    fill_column = log2FoldChange,
    fill_type = "fold_change",
    layout = "nicely",
    label = TRUE,
    label_column = gene_name,
    label_filter = 5,
    legend = TRUE
)
```


# Performing pathway enrichment
The essence of pathway enrichment is the concept of over-representation. In
other words, are there more "pathway A" genes in a list of analyzed genes (e.g.,
DEGs) than expected? To calculate this, the simplest method is to compare the
ratio of DEGs in pathway A to all DEGs and all genes in pathway A to all genes
in all pathways in a database, which is termed "over-representation analysis".
In this package, we primarily use the Reactome database (Fabregat et al. 2017),
although we may incorporate others (KEGG, GO) in the future.

One issue that occurs with over-representation analysis is the assumption that
each gene in each pathway has "equal" value and belonging to that pathway. In
reality, a single protein can have multiple (and sometimes very different)
functions and belong to multiple pathways, like protein kinases. There are also
pathways that have substantial overlap of cellular machinery, like the TLR
pathways. This can lead to enrichment of multiple similar pathways or even
unrelated "false-positive" pathways that make parsing through results very
difficult.

One solution is to use gene pairs, as elegantly described by the creators of the
package `sigora` (Foroushani et al. 2013). In our experience, this methodology
decreases the number of similar pathways as well as unrelated pathways from
promiscuous genes, leaving behind only the pathways that are likely related to
the underlying biology. In our pathway enrichment function `enrich_pathway()`,
the default enrichment method is `sigora`.

The `enrich_pathway()` function takes in a list of DESeq2 result data frames,
and by default will split the genes into up and down-regulated genes, then
perform enrichment separately (set `split = FALSE` to perform enrichment without
splitting). The name of the dataframe(s) in the list should indicate the
comparison that resulted in the DESeq2 results.
```{r sigora_enrichment}
# Perform enrichment with Sigora (default). We also need to specify the
# gene-pair signature repository (more in the Supplemental section).
enriched_results_sigora <- enrich_pathway(
    input_list = deseq_example_list,
    gps_repo = reaH
)

# Enrichment without splitting by fold change direction
enriched_results_sigora_nosplit <- enrich_pathway(
    input_list = deseq_example_list,
    gps_repo = reaH, 
    split = FALSE
)
```

For those who still prefer traditional over-representation analysis, we include
the option of doing so by setting `analysis = "reactomepa"`, which uses
ReactomePA (Yu et al. 2016).
```{r reactomepa_enrichment}
# Perform enrichment with ReactomePA
enriched_results_rpa <- enrich_pathway(
    input_list = deseq_example_list,
    analysis = "reactomepa"
)
```

In addition to the Reactome database, we also provide over-representation
analysis using the Hallmark Gene Signatures from the Molecular Signatures
Database. These are 50 gene sets that represent “specific, well-defined
biological states or processes with coherent expression” (Liberzon et al. 2015).
This database provides a more high-level summary of key biological processes
compared to the more granular Reactome pathways.
```{r hallmark_enrichment}
enriched_results_hm <- enrich_pathway(
    input_list = deseq_example_list,
    analysis = "hallmark"
)
```


# Plotting pathway enrichment results
Now that we have (a lot of) pathway enrichment results from multiple
comparisons, it is time to visualize them. The function `plot_pathways` does
this by grouping Reactome pathways or Hallmark gene sets under parent groups and
indicating if each pathway is up/down-regulated across different comparisons,
making it easy to identify which pathways are shared or unique to which
comparisons. Because there are often many pathways, you can split the plot up
into multiple columns (up to 3) or truncate the pathway names.

Sometimes, after splitting, a pathway can be enriched by both up and
down-regulated genes (these usually occur with larger pathways), and are
indicated by a white asterisk where the more significant (lower p-value)
directionality is displayed. You can also change the angle/labels of the
comparisons or add the number of DEGs in each comparison to the labels. Lastly,
you can specify which pathways or top pathway groups to pull out to visualize.

```{r plot_pathways_I, fig.width=12, fig.height=10}
# Plot the Sigora results in two columns
plot_pathways(enriched_results_sigora, columns = 2)
```

```{r plot_pathways_II, fig.width=8, fig.height=8}
# Plot the Sigora results with the following tweaks:
# - Only the immune system pathways
# - Condense the names
# - Add the number of DEGs used for enrichment
# - Make the labels horizontal
# - Increase the truncation cutoff value for more words
plot_pathways(
    enriched_results_sigora, 
    specific_top_pathways = "Immune System",
    new_group_names = c("Pos", "Neg"), 
    show_num_genes = TRUE,
    x_angle = "horizontal",
    name_width = 50
)
```

From these results, you can see that while many of the immune system pathways
change in the same direction over time in COVID-19 and non-COVID-19 sepsis
patients, a few unique ones stand out, mostly related to interferon signaling
("Interferon Signaling", "Interferon gamma signaling", "Interferon alpha/beta
signaling", "ISG15 antiviral mechanism"). This likely reflects an elevated early
antiviral response in COVID-19 patients that decreased over time, compared to no
change in non-COVID-19 sepsis patients.


# Generating networks from enriched pathways
`pathnet` also includes functions for turning the pathway enrichment results
from `sigora` into networks, using the overlap of the genes assigned to each
pathway to determine their similarity to each other. In these networks, each
pathway is a node, with connections or edges between them determined via
distance measures. A threshold can be set, where two pathways with a minimum
similarity measure are considered connected, and would have an edge drawn
between their nodes.

We provide a pre-computed distance matrix of Reactome pathways, generated using
Jaccard distance, but there is support for multiple distance measures to be
used.
```{r pathway_network_I, fig.width=12, fig.height=10, warning=FALSE}
starting_pathways <- create_foundation(
    mat = pathway_distances_jaccard,
    max_distance = 0.8
)

ex_pathway_network_input <- enriched_results_sigora %>% 
    filter(comparison == "COVID Pos Over Time")

my_pathway_network <- create_pathnet(
    sigora_result = sigora_examples,
    foundation = starting_pathways,
    trim = TRUE,
    trim_order = 1
)

# Static plot with ggraph
pathnet_ggraph(
    my_pathway_network,
    label_prop = 0.1,
    node_label_size = 4,
    node_label_overlaps = 8,
    seg_colour = "red"
)
```

```{r pathway_network_II}
# Interactive plot with visNetwork
pathnet_visNetwork(my_pathway_network)
```

## Pathway networks using only DE genes
```{r pathway_network_III}
candidate_data <- sigora_examples %>%
    filter(comparison == "COVID Pos Over Time") %>% 
    select(pathway_id, genes) %>%
    separate_rows(genes, sep = ";") %>%
    left_join(
        sigora_database,
        by = c("pathway_id", "genes" = "Symbol"),
        multiple = "all"
    ) %>%
    relocate(pathway_id, Ensembl.Gene.ID, "Symbol" = genes, pathway_name) %>%
    distinct()

# Now that we have a smaller table in the same format as sigora_database, we can
# construct our own matrix of pathway distances
candidate_dist_data <- get_pathway_distances(
    pathway_data = candidate_data,
    dist_method = "jaccard"
)

candidate_starting_pathways <- create_foundation(
    mat = candidate_dist_data,
    max_distance = 0.9
)

candidates_as_network <- create_pathnet(
    sigora_result = filter(sigora_examples, comparison == "COVID Pos Over Time"),
    foundation = candidate_starting_pathways,
    trim = FALSE
)

pathnet_visNetwork(candidates_as_network)
```


# Visualizing fold changes across comparisons
We can also verify this by looking at fold change heatmaps of genes involved in
these pathways. The function `plot_fold_change` can do so by taking in an input
list of DESeq2 data frames.
```{r plot_fold_changes_I, fig.height=10}
plot_fold_change(
    deseq_example_list,
    path_name = "Interferon alpha/beta signaling",
    col_split = c("Positive", "Negative"),
    cluster_columns = FALSE
)
```

Another example, with more customization:
```{r plot_fold_changes_II}
deseq_example_list_renamed <- purrr::set_names(
    deseq_example_list,
    c("Pos", "Neg")
)

plot_fold_change(
    deseq_example_list_renamed,
    manual_title = "Genes I like",
    genes_to_plot = c("CD4", "CD8A","CD8B", "CD28", "ZAP70"),
    gene_format = "hgnc",
    col_split = c("Pos", "Neg"),
    log2_foldchange = TRUE,
    col_angle = 45,
    cluster_rows = FALSE,
    cluster_columns = TRUE,
    invert = TRUE
)
```


# Supplemental materials

## Gene-pair signatures
`sigora` uses gene-pair signature (GPS) repositories that stores information on
which gene pairs are unique for which pathways. We recommend using the one
already loaded with `sigora`, which is `reaH` (Reactome Human). You can also
generate your own GPS repo using `sigora` and your own set of pathways (e.g.
from another pathway database like GO or KEGG). Please consult `sigora`
documentation on how to generate your custom GPS repo.

## Why are there different p-value cut-offs for Sigora vs ReactomePA/Hallmark?
Because there are now multiple gene pairs vs just genes, the gene-pair
"universe" is greatly increased and it is more likely for a result to be
significant. Therefore, the cutoff threshold for significance is more stringent
(p-value < 0.001) and a more conservative adjustment method (Bonferroni) is
used. For regular over-representation analysis, a less conservative adjustment
method (Benjamini-Hochberg) is used with adjusted p-value <0.05. These are
automatically set with `filter_results = "default"`. You can adjust these
cut-offs by setting `filter_results` to different values between 0 and 1, or `1`
if you want all the pathways (this may be useful for comparing which enriched
genes appear in which comparisons, even if the enrichment is not significant).


# Citations
An AY, Baghela AS, Falsafi R, Lee AH, Trahtemberg U, Baker AJ, dos Santos CC,
Hancock REW. Severe COVID-19 and non-COVID-19 severe sepsis converge
transcriptionally after a week in the intensive care unit, indicating common
disease mechanisms. Front Immunol. 2023;6(14):1167917.

Fabregat A, Sidiropoulos K, Viteri G, Forner O, Marin-Garcia P, Arnau V,
D’Eustachio P, Stein L, Hermjakob H. Reactome pathway analysis: a
high-performance in-memory approach. BMC Bioinform. 2017;18:142.

Foroushani ABK, Brinkman FSL, Lynn DJ. Pathway-GPS and Sigora: identifying
relevant pathways based on the over-representation of their gene-pair
signatures. PeerJ. 2013;1:e229.

Liberzon A, Birger C, Thorvaldsdóttir H, Ghandi M, Mesirov JP, Tamayo P. The
Molecular Signatures Database (MSigDB) hallmark gene set collection. Cell Syst.
2015;1(6):417–25.

Love MI, Huber W, Anders S. Moderated estimation of fold change and dispersion
for RNA-seq data with DESeq2. Genome Biol. 2014;15(12):550.

Yu G, He QY. ReactomePA: an R/Bioconductor package for reactome pathway analysis
and visualization. Mol Biosyst. 2016;12(2):477-9.


# Session information
```{r session_information, echo=FALSE}
sessionInfo()
```
